parquet/src/main/java/org/apache/iceberg/parquet/ParquetValueWriters.java

```diff
@@ -648,7 +656,7 @@ protected StructWriter(List<ParquetValueWriter<?>> writers) {
     @Override
     public void write(int repetitionLevel, S value) {
       for (int i = 0; i < writers.length; i += 1) {
-        Object fieldValue = get(value, i);
+        Object fieldValue = get(value, fieldIndexes[i]);
```

Comment:
The fix is to map from writer index to field index. The index is created below by skipping unknown fields in the data struct.